Enhance Open SCAP tests and add more profiles tests

vtrubovics opened the pull request:
Enhance Open SCAP tests and add more profiles bash and ansible tests:
xccdf_org.ssgproject.content_profile_stig
xccdf_org.ssgproject.content_profile_cis
xccdf_org.ssgproject.content_profile_pci-dss-4
xccdf_org.ssgproject.content_profile_hipaa
xccdf_org.ssgproject.content_profile_anssi_bp28_high

External configuration, exclusions and expected results.
DS and ansible file modifications.
Better results analysis and display.
Results collection to log file.
Ansible playbook enhanced run and result analysis.

- Related ticket: https://progress.opensuse.org/issues/150863
- Verification runs: 
#x86 SP4 POO150863
security/oscap_anssi_bp_28_high.yaml https://openqa.suse.de/tests/12811263
security/oscap_hipaa.yaml https://openqa.suse.de/tests/12811264
security/oscap_cis.yaml https://openqa.suse.de/tests/12811265
security/oscap_pci_dss_4.yaml https://openqa.suse.de/tests/12811266
security/oscap_stig.yaml https://openqa.suse.de/tests/12811267
security/oscap_ansible_pci_dss_4.yaml https://openqa.suse.de/tests/12811268
security/oscap_ansible_anssi_bp_28_high.yaml https://openqa.suse.de/tests/12811269
security/oscap_ansible_hipaa.yaml https://openqa.suse.de/tests/12811270
security/oscap_ansible_stig.yaml https://openqa.suse.de/tests/12811271
security/oscap_ansible_cis.yaml https://openqa.suse.de/tests/12811272

#x86 SP5 POO150863
oscap_anssi_bp_28_high.yaml https://openqa.suse.de/tests/12811273
oscap_hipaa.yaml https://openqa.suse.de/tests/12811274
oscap_cis.yaml https://openqa.suse.de/tests/12811275
oscap_pci_dss_4.yaml https://openqa.suse.de/tests/12811276
oscap_stig.yaml https://openqa.suse.de/tests/12811277
oscap_ansible_pci_dss_4.yaml https://openqa.suse.de/tests/12811278
oscap_ansible_anssi_bp_28_high.yaml https://openqa.suse.de/tests/12811279
oscap_ansible_hipaa.yaml https://openqa.suse.de/tests/12811280
oscap_ansible_stig.yaml https://openqa.suse.de/tests/12811281
oscap_ansible_cis.yaml https://openqa.suse.de/tests/12811282

#ARM SP4 POO150863
oscap_anssi_bp_28_high.yaml https://openqa.suse.de/tests/12811283
oscap_hipaa.yaml https://openqa.suse.de/tests/12811284
oscap_cis.yaml https://openqa.suse.de/tests/12811285
oscap_pci_dss_4.yaml https://openqa.suse.de/tests/12811286
oscap_stig.yaml https://openqa.suse.de/tests/12811287
oscap_ansible_pci_dss_4.yaml https://openqa.suse.de/tests/12811288
oscap_ansible_anssi_bp_28_high.yaml https://openqa.suse.de/tests/12811289
oscap_ansible_hipaa.yaml https://openqa.suse.de/tests/12811291
oscap_ansible_stig.yaml https://openqa.suse.de/tests/12811297
oscap_ansible_cis.yaml https://openqa.suse.de/tests/12811304

#ARM SP5 POO150863
oscap_anssi_bp_28_high.yaml https://openqa.suse.de/tests/12811343
oscap_hipaa.yaml https://openqa.suse.de/tests/12811344
oscap_cis.yaml https://openqa.suse.de/tests/12811345
oscap_pci_dss_4.yaml https://openqa.suse.de/tests/12811346
oscap_stig.yaml https://openqa.suse.de/tests/12811347
oscap_ansible_pci_dss_4.yaml https://openqa.suse.de/tests/12811348
oscap_ansible_anssi_bp_28_high.yaml https://openqa.suse.de/tests/12811349
oscap_ansible_hipaa.yaml https://openqa.suse.de/tests/12811350
oscap_ansible_stig.yaml https://openqa.suse.de/tests/12811351
oscap_ansible_cis.yaml https://openqa.suse.de/tests/12811352

#s390x SP4 POO150863
oscap_anssi_bp_28_high.yaml https://openqa.suse.de/tests/12811353
oscap_hipaa.yaml https://openqa.suse.de/tests/12811354
oscap_cis.yaml https://openqa.suse.de/tests/12811355
oscap_pci_dss_4.yaml https://openqa.suse.de/tests/12811356
oscap_stig.yaml https://openqa.suse.de/tests/12811357
oscap_ansible_pci_dss_4.yaml https://openqa.suse.de/tests/12811358
oscap_ansible_anssi_bp_28_high.yaml https://openqa.suse.de/tests/12811363
oscap_ansible_hipaa.yaml https://openqa.suse.de/tests/12811366
oscap_ansible_stig.yaml https://openqa.suse.de/tests/12811368
oscap_ansible_cis.yaml https://openqa.suse.de/tests/12811369

#s390x SP5 POO150863
oscap_anssi_bp_28_high.yaml https://openqa.suse.de/tests/12811373
oscap_hipaa.yaml https://openqa.suse.de/tests/12811374
oscap_cis.yaml https://openqa.suse.de/tests/12811375
oscap_pci_dss_4.yaml https://openqa.suse.de/tests/12811376
oscap_stig.yaml https://openqa.suse.de/tests/12811377
oscap_ansible_pci_dss_4.yaml https://openqa.suse.de/tests/12811378
oscap_ansible_anssi_bp_28_high.yaml https://openqa.suse.de/tests/12811379
oscap_ansible_hipaa.yaml https://openqa.suse.de/tests/12811380
oscap_ansible_stig.yaml https://openqa.suse.de/tests/12811381
oscap_ansible_cis.yaml https://openqa.suse.de/tests/12811382

jknphy commented:
Please squash all the commits in single one or split it in commits that has their own goal and description if apply.
Then would be informative to have the two first paragraph of the PR's 1st comment included in the resulting commit(s).

jknphy commented on tests/security/oscap_ansible_anssi_bp_28_high/oscap_security_guide_setup.pm:
```diff
@@ -0,0 +1,31 @@
+    $oscap_tests::sle_version = 'sle' . get_required_var('VERSION') =~ s/([0-9]+).*/$1/r;
+    $oscap_tests::evaluate_count = 3;
+    $oscap_tests::ansible_remediation = 1;
+    $oscap_tests::profile_ID = is_sle ? $oscap_tests::profile_ID_sle_anssi_bp28_high : $oscap_tests::profile_ID_tw;
```
This should be a setting, we are repeating this test, for example:
`tests/security/oscap_ansible_stig/oscap_security_guide_setup.pm`
Same for the line 21, it could be parametrized and would save a lot of files.

Additionally I noticed you missed to select the console in some of these files more below.

vtrubovics replied:
Fixed

jknphy commented:
It would also be more clear if instead of having so many links in your PR's 1st comment, you would just provide 1 link with the whole matrix, when you use a tool like `openqa-clone-custom-git-refspec` it creates a branch for you when you can see everything and even to refine the url to show more products, it would avoid for you to go to links and edit everything for a change or re-run (if you have some initial wrong test suites, you might need to delete them and clean them up, but other than that inconvenient it is the standard way to provide test matrix).

jknphy commented on tests/security/oscap_pci_dss_4/oscap_xccdf_remediate.pm:
```diff
@@ -0,0 +1,25 @@
+sub run {
+    my ($self) = @_;
+
+
```
neatpicking: avoid more than one blank like to separate code blocks.

vtrubovics replied:
File will be deleted during refactoring.

jknphy commented on tests/security/oscap_pci_dss_4/oscap_xccdf_remediate.pm:
```diff
@@ -0,0 +1,25 @@
+# Copyright 2022 SUSE LLC
```
s/2022/2023/g

jknphy commented on tests/security/oscap_pci_dss_4/oscap_xccdf_eval_remote.pm:
```diff
@@ -0,0 +1,26 @@
+use warnings;
+use testapi;
+use utils;
+use version_utils qw(is_sle);
```
please double-check for libraries not used in all files (unfortunately Perl or the IDE doesn't help us much here).

vtrubovics replied:
Checked

jknphy commented on tests/security/oscap_pci_dss_4/oscap_xccdf_remediate.pm:
```diff
@@ -0,0 +1,25 @@
+#
+# Summary: Test 'pci-dss-4' hardening in the 'scap-security-guide': mitigation mode
+# Maintainer: QE Security <none@suse.de>
+# Tags: poo#93886, poo#104943
```
This Tags field stopped to make sense a lot of time ago, unless it is really specific for this, feel free to remove it, specially when more than one ticket is added is one start to accumulate and being less usuful.

vtrubovics replied:
Fixed

jknphy commented on lib/oscap_tests.pm:
```diff
@@ -123,186 +711,541 @@ sub upload_logs_reports {
 
+sub oscap_security_guide_setup {
+    # Main test setup function
     select_console 'root-console';
```
this is repeated, the console is selected twice, from the test and here, the library should be agnostic of the type console used, could be outside in the test layer.

vtrubovics replied:
Fixed

jknphy commented on lib/oscap_tests.pm:
```diff
@@ -41,53 +59,367 @@ our $f_vlevel = 'ERROR';
+    $f_ssg_sle_xccdf =
+      '/usr/share/xml/scap/ssg/content/ssg-sle' . "$version" . '-xccdf.xml';
+}
+sub set_ds_file_name {
```
missing separation between functions.

vtrubovics replied:
Fixed

jknphy commented on lib/oscap_tests.pm:
```diff
@@ -41,53 +59,367 @@ our $f_vlevel = 'ERROR';
+    }
+}
 
+sub replace_ansible_file {
```
functions in this library are extremely long and might need proper function documentation, Perldoc, there are multiple example in the repo, but I would recommend to leave this comment open and do it once all review regarding the library itself is concluded to avoid extra work, because we could consider split some functions but I need to take a closer look...

jknphy commented on lib/oscap_tests.pm:
```diff
@@ -123,186 +711,541 @@ sub upload_logs_reports {
+sub oscap_security_guide_setup {
+    # Main test setup function
     select_console 'root-console';
+    # Setting $full_ansible_file_path aftr got ansible_profile_ID from test
```
better not introduce names of variables in the comments, because variable names can changes in the future and the comments might be outdated easily.
Additionally I would like to mention that we don't have any code style for having a comment for each block or line, normally a comment is only needed if there is an exception or something that is not obvious by the code or require some experience in the area. For example this comment is obvious, it is not needed, similar above for `#Main setup function` which is attached to the console selection, that should be out, same than for a `zypper in` and you wil find more example when reading the code. It helps to readability as well to have comments only where they are needed.

vtrubovics replied:
I am writing code and comments as much as possible commented and understandable.
Will cleanup possibly unneeded comments.

jknphy commented on lib/oscap_tests.pm:
```diff
@@ -41,53 +59,367 @@ our $f_vlevel = 'ERROR';
     # for SLE15 the ds file is "ssg-sle15-ds.xml";
     # for SLE12 the ds file is "ssg-sle12-ds.xml";
     # for Tumbleweed the ds file is "ssg-opensuse-ds.xml"
     my $version = get_required_var('VERSION') =~ s/([0-9]+).*/$1/r;
```
instead of running this regex in multiple places looks like these two methods could be replaced with openQA variables and be fixed there as well.

jknphy commented on lib/oscap_tests.pm:
```diff
@@ -41,53 +59,367 @@ our $f_vlevel = 'ERROR';
+our $sle_version = '';
 
 # Upload HTML report by default
 set_var('UPLOAD_REPORT_HTML', 1);
```
unless you need to change this during the test for some very specific reason, if it is fixed, should be out in openQA variables as a flag that can be tweaked, or perhaps just a normal Perl variable set if doesn't change.

vtrubovics replied:
moved $test_run_report_name to local var

jknphy commented on lib/oscap_tests.pm:
```diff
@@ -41,53 +59,367 @@ our $f_vlevel = 'ERROR';
 
+sub replace_ds_file {
+    # Replace original ds file whith built or downloaded from repository
+    my $self = $_[0];
```
it is more typical to use in Perl ` my ($self) = @_`

vtrubovics replied:
Fixed

jknphy commented on lib/oscap_tests.pm:
```diff
@@ -123,186 +711,541 @@ sub upload_logs_reports {
     $out = script_output("oscap -V");
     record_info("oscap version", "\"# oscap -V\" returns:\n $out");
+
+    # Get the tests configuration file from repository and set global configuration varables
```
I noticed multiple typos in the comments, `varables', below 'whith', 'depencies' ans so on, as mentioned in another comment, better to avoid comments when there is obvious code, they attract this kind of problems. Please give a double-check to them and remove all that you find superfluous.

vtrubovics replied:
Fixed

jknphy commented on lib/oscap_tests.pm:
```diff
@@ -123,186 +711,541 @@ sub upload_logs_reports {
+sub get_tests_config {
+    # Get the tests configuration file from repository
+    my $config_file_name = "openqa_config.conf";
+    my $url = "https://gitlab.suse.de/seccert-public/compliance-as-code-compiled/-/raw/main/content/";
```
we are revealing the information of this file somehow all over the places in comments in this PR so I guess the file we are fetching here is not under some security concern even if it is on that repo, I'm wondering if we could simplify it and just have this file stored in this repo under data/ instead of fetching it. It might get out of sync, but doesn't look like it has much information... wdyt?

Note: we usually use [get_test_data](https://open.qa/api/testapi/#_get_test_data) for fetching from data/ folder.

vtrubovics replied:
It is complicated because:
1. Improvement work constantly going on by Hardening team for multiple profiles and results/profiles  are constantly changing.
2. Every 3 months there are OSCAP release and results will change.
3. There are need to test profiles from different sources.
For this reason I made gitlab repository where OSCAP under test, expected results, configuration and exclusions are stored.
So we can do changes there quickly without PRs.
When SCAP profiles will stabilize - we will consider to move config part to openQA.

jknphy commented on lib/oscap_tests.pm:
```diff
@@ -123,186 +711,541 @@ sub upload_logs_reports {
+    my $config = Config::Tiny->new;
+    $config = Config::Tiny->read_string("$data");
+    my $err = Config::Tiny::errstr;
+    if ($err eq "") {
```
skip the error handling, if it fails it should die and a stack trace is printed anyway in openQA, so I think it is not needed, also notice typo `s/returened/returned/`.
Additionally I think the return value is not used anywhere, but if you pick this from data/ you would not even need this function.

vtrubovics replied:
Fixed typo.
I think error handling is valid here.
In case of phrasing error it will use default values.

jknphy commented on lib/oscap_tests.pm:
```diff
@@ -123,186 +711,541 @@ sub upload_logs_reports {
+    # Get the tests configuration file from repository and set global configuration varables
+    get_tests_config();
+    push(@test_run_report, "[configuration]");
+    $out = script_output("date", quiet => 1);
```
In general in Perl is good to avoid doing too many assignations, one would end up controlling too many variables and it is easy to mess up, you would see more frequent that his operation is done in-line concatenating directly in the next line 879. Make it more readable and also you don't have to control mentally a reused $out variable, this kind of flow sometimes might end up in getting some debug from $out which was not from the step you wanted and the reader getting a bit confused.

vtrubovics replied:
Fixed

jknphy commented on lib/oscap_tests.pm:
```diff
@@ -123,186 +711,541 @@ sub upload_logs_reports {
+    get_tests_config();
+    push(@test_run_report, "[configuration]");
+    $out = script_output("date", quiet => 1);
+    push(@test_run_report, "date = $out");
+    push(@test_run_report, "profile_ID = $profile_ID");
+    push(@test_run_report, "ansible_profile_file_name = $ansible_profile_ID");
+    push(@test_run_report, "use_content_type = $use_content_type");
+    push(@test_run_report, "remove_rules_missing_fixes = $remove_rules_missing_fixes");
+    push(@test_run_report, "use_exclusions = $use_exclusions");
+    push(@test_run_report, "evaluate_count = $evaluate_count");
+
+    # Replace original ds and xccdf files whith downloaded from local repository
+    set_ds_file_name();
+    push(@test_run_report, "sle_version = $sle_version");
+    push(@test_run_report, "ssg_sle_ds = $ssg_sle_ds");
+    push(@test_run_report, "ssg_sle_xccdf = $ssg_sle_xccdf");
+    my $arch = get_var 'ARCH';
+    push(@test_run_report, "os_arch = $arch");
+    my $type;
+    if ($ansible_remediation == 1) {
+        $type = 'ansible';
+    }
+    else {
+        $type = 'bash';
+    }
+    push(@test_run_report, "remediation_type = $type");
+    my $build_url = get_var 'CASEDIR';
+    push(@test_run_report, "build_url = $build_url");
+    my $iso_name = get_var 'ISO';
+    push(@test_run_report, "iso_name = $iso_name");
+    my $full_name = get_var 'NAME';
+    push(@test_run_report, "full_name = $full_name");
+    my $schedule = get_var 'YAML_SCHEDULE';
+    push(@test_run_report, "schedule = $schedule");
```
all this block could be extracted to a function called for example `create_run_report`

vtrubovics replied:
The reason why I did not create create_run_report function - data collection done in multiple functions and in different ways, code never reused.

jknphy commented on lib/oscap_tests.pm:
```diff
@@ -123,186 +711,541 @@ sub upload_logs_reports {
+        # Need to use pyython3.1x
+        add_suseconnect_product(get_addon_fullname('python3'));
+        # On SLES 12 ansible packages require depencies located in sle-module-public-cloud
+        add_suseconnect_product(get_addon_fullname('pcm'), (is_sle('<15') ? '12' : undef)) if is_sle('<15');
```
this is checking twice that is <15.

vtrubovics replied:
Fixed

jknphy commented on lib/oscap_tests.pm:
```diff
@@ -123,186 +711,541 @@ sub upload_logs_reports {
+
+    # If required ansible remediation
+    if ($ansible_remediation == 1) {
+        my $pkgs = 'ansible';
```
same here, better not create those intermediate assignations, just concatenate the string, also $pkgs is plural, which is not expected.

vtrubovics replied:
Fixed var name.
So how it is better to do assignations?

jknphy commented on lib/oscap_tests.pm:
```diff
@@ -123,186 +711,541 @@ sub upload_logs_reports {
+    if ($ansible_remediation == 1) {
+        my $pkgs = 'ansible';
+        # Need to update SLES to fix issues with STIG playbook
+        record_info("Update", "Updaiting SLES");
```
s/Updaiting/Updating/

vtrubovics replied:
Fixed. Removed update call, as it is not needed.

jknphy commented on lib/oscap_tests.pm:
```diff
@@ -123,186 +711,541 @@ sub upload_logs_reports {
+        #install ansible.posix
+        assert_script_run("ansible-galaxy collection install ansible.posix");
+    }
+    if ($remove_rules_missing_fixes == 1 or $use_content_type == 3) {
```
I'm not that expert in Perl, but we always (or most of the time in the repo) use `||` instead of `or` I think that is because of the precedence, that avoid some tricky mistakes.

vtrubovics replied:
Can use both, but "or" has low precedence: 
!, &&, ||, and ^ have high precedence such that they are useful in constructing an expression; not, and, or, and xor have low precedence such that they are useful for flow control between what are essentially different expressions.

Fixed it to:
(($use_content_type == 2) or ($use_content_type == 3))

jknphy commented on lib/oscap_tests.pm:
```diff
@@ -123,186 +711,541 @@ sub upload_logs_reports {
+            replace_ansible_file();
+        }
+    }
+    # Adding benchmark version to report
```
here again missing blank line, the comment doesn't count as separation.

vtrubovics replied:
Fixed

jknphy commented on lib/oscap_tests.pm:
```diff
@@ -123,186 +711,541 @@ sub upload_logs_reports {
+        # Get bash and ansible rules lists from data based on provided
+        my $ansible_rules_missing_fixes_ref;
+        my $bash_rules_missing_fixes_ref;
+        modify_ds_ansible_files($mising_rules_full_path, $bash_rules_missing_fixes_ref, $ansible_rules_missing_fixes_ref);
```
the name of the function should reflect also that does something with bash.

vtrubovics replied:
What do you mean?
modify_ds_ansible_files returns array references to bash_rules_missing_fixes array and ansible_rules_missing_fixes array.

jknphy commented on lib/oscap_tests.pm:
```diff
@@ -123,186 +711,541 @@ sub upload_logs_reports {
+        $found = 1;
+        if (not defined $$eval_match[0]) {
+            my @eval_match = ();
+            $_[0] = \@eval_match;
```
this afir in Perl is not recommended, you are changing the input parameters, I tried to follow from the test the expected result and was really complicated.
For this you can google it but I could find for example [this](https://stackoverflow.blog/2022/09/08/this-is-not-your-grandfathers-perl/)

vtrubovics replied:
Fixed - Simplified data file and functions get_test_exclusions, get_test_expected_results

jknphy commented on lib/oscap_tests.pm:
```diff
@@ -41,53 +59,367 @@ our $f_vlevel = 'ERROR';
+        record_info("Copied ansible file", "Copied file $ansible_profile_ID to $full_ansible_file_path");
+    }
+    # scap-security-guide
+    elsif ($use_content_type == 1) {
```
instead of extensive comments, in general in the code some contants could be created on the top of the file in capital letters to represent those numeric values, kind of like enums in other languagues, leaving 0s and 1s like clear booleans.

jknphy commented on tests/security/oscap_ansible_anssi_bp_28_high/oscap_xccdf_eval.pm:
```diff
@@ -0,0 +1,29 @@
+    select_console 'root-console';
+
+    # Set expected results
+    my @eval_match = ('');
```
looks really hacky to pass this, most likely you want to set default param in your function

vtrubovics replied:
Deleted file

jknphy commented on lib/oscap_tests.pm:
```diff
@@ -123,186 +711,541 @@ sub upload_logs_reports {
     select_console 'root-console';
 
+    my $n_failed_rules = @$eval_match;
+    my ($failed_rules_ref, $passed_rules_ref);
```
feels in general in the code that you are missing intermediate structure to store values, in this case doesn't make sense to save `my` to indicate that are together, anyway are not connected variables. Introductions of hashes with well-named keys could help to read the code and connect parts that are related.

jknphy commented on lib/oscap_tests.pm:
```diff
@@ -123,186 +711,541 @@ sub upload_logs_reports {
-    if ($ret != 0 and $ret != 2) {
-        record_info('bsc#1194676', 'remediation should be succeeded', result => 'fail');
+    # For pkg install rules need to refresh repositories
+    # zypper_call('ref -s', timeout => 180);
```
commented code that should be removed if not executed.

vtrubovics replied:
Cleaned up code

jknphy commented on lib/oscap_tests.pm:
```diff
@@ -13,21 +13,39 @@ use warnings;
 use testapi;
```
normally we put this imports after the base or parent, which is usually the first one.

vtrubovics replied:
Fixed

jknphy commented on lib/oscap_tests.pm:
```diff
@@ -41,53 +59,367 @@ our $f_vlevel = 'ERROR';
+    elsif ($use_content_type == 2) {
+        download_file_from_https_repo($url, $xccdf_file_name);
+        # Remove original xccdf file
+        assert_script_run("rm $f_ssg_sle_xccdf");
+        # Copy downloaded file to correct location
+        assert_script_run("cp $xccdf_file_name $f_ssg_sle_xccdf");
+        record_info("Copied xccdf file", "Copied file $xccdf_file_name to $f_ssg_sle_xccdf");
```
these 3 actions could run after the if/elseif and you just set the values in the branches.

vtrubovics replied:
Not the case here: in each  if/elseif there are different cp source.

jknphy commented on lib/oscap_tests.pm:
```diff
@@ -41,53 +59,367 @@ our $f_vlevel = 'ERROR';
+sub backup_ds_file {
+    # Backup ds file for later reuse
+    assert_script_run("cp $f_ssg_ds /root/$ssg_sle_ds");
+    record_info("Backuped ds file", "Backuped file $f_ssg_ds to /root/$ssg_sle_ds");
```
s/Backuped/Backed up/

vtrubovics replied:
Fixed

jknphy commented on lib/oscap_tests.pm:
```diff
@@ -98,14 +430,270 @@ sub upload_logs_reports {
+    $_[2] = \@ansible_rules;
+}
+
+sub generate_mising_rules {
```
s/mising/missing/

vtrubovics replied:
Fixed

jknphy commented:
I think this is ok for some 1st review round. In summary:
1. Too many similar files (test and yamls)-> using openQA settings should remediate that.
2. One big library file (more than 1k lines) that should be split in small functions or even files and more readable code (the use of global variables doesn't help, function with clear parameters always does).

I'm wondering if it would be better to split this PR in multiple ones, to see single code path, for example focusing on one test, then it would be more feasible to review the library, otherwise it is very tricky, too many cases. But let's see how it looks after addressing the comments you think worth to address.

vtrubovics commented on lib/oscap_tests.pm:
```diff
@@ -123,186 +711,541 @@ sub upload_logs_reports {
-    my $f_ssg_ds = is_sle ? $f_ssg_sle_ds : $f_ssg_tw_ds;
-    $out = script_output("oscap info $f_ssg_ds");
+    $f_ssg_ds = is_sle ? $f_ssg_sle_ds : $f_ssg_tw_ds;
+    $out = script_output("oscap info $f_ssg_ds", quiet => 1);
     record_info("oscap info", "\"# oscap info $f_ssg_ds\" returns:\n $out");
-
     # Check the oscap version information for reference
     $out = script_output("oscap -V");
     record_info("oscap version", "\"# oscap -V\" returns:\n $out");
+
```
fixed

vtrubovics commented on lib/oscap_tests.pm:
```diff
@@ -123,186 +711,541 @@ sub upload_logs_reports {
+        push(@test_run_report, "[tests_results]");
     }
+    # If doing ansible playbook remediation
+    if ($ansible_remediation == 1) {
```
Do not agree here. oscap_remediate function is complex because of multiple reasons:
1. 2 types of remediations
2. Complex ansible remediation (2 sequential runs) and additional analysis in case of unsuccessful ansible playbook execution.
3. Most of analysis moved out to functions:   ansible_result_analysis, ansible_failed_tasks_search_vv, find_ansible_cce_by_task_name_vv

vtrubovics commented:
> I think this is ok for some 1st review round. In summary:
> 
>     1. Too many similar files (test and yamls)-> using openQA settings should remediate that.
> 
>     2. One big library file (more than 1k lines) that should be split in small functions or even files and more readable code (the use of global variables doesn't help, function with clear parameters always does).
> 
> 
> I'm wondering if it would be better to split this PR in multiple ones, to see single code path, for example focusing on one test, then it would be more feasible to review the library, otherwise it is very tricky, too many cases. But let's see how it looks after addressing the comments you think worth to address.

I trying to solve complex problem of testing  SCAP profiles inside OpenQA. SCAP itself is quite big framework with big functionality.
I tried different approaches to implement required required (from hardening team) functionality with minimal complexity and good flexibility. 
These tests will evolve with additional new requirements and remediation methods.

vtrubovics commented on lib/oscap_tests.pm:
```diff
@@ -41,53 +59,368 @@ our $f_vlevel = 'ERROR';
+our $test_run_report_name = "test_run_report.txt";
+
+# Get sle version "sle12" or "sle15"
+our $sle_version = 'sle' . get_required_var('VERSION') =~ s/([0-9]+).*/$1/r;
```
Fixed

tjyrinki commented:
Superseded by https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/18503 which should be the same but with conflicts resolved.
